### PR TITLE
Configurations and DeviceConfigurations

### PR DESCRIPTION
**Pull request by jaredmorrow**
Adds support for the following endpoints:

```csv
GET	/configuration
POST	/configuration
GET	/configuration/:id
GET	/configuration/:id/device-configuration
GET	/configuration/:id/relationships/device-configuration
GET	/device-configuration
POST	/device-configuration
GET	/device-configuration/:id
DELETE	/device-configuration/:id
GET	/device-configuration/:id/configuration
GET	/device-configuration/:id/device
GET	/device-configuration/:id/relationships/configuration
GET	/device-configuration/:id/relationships/device
GET	/sensor/:id/device-configuration
GET	/sensor/:id/relationships/device-configuration
GET	/element/:id/device-configuration
GET	/element/:id/relationships/device-configuration
```

Relationships can be used any direction given how Configurations / Device Configurations / Devices interact.

Rules:
1. Configurations are immutable, they **cannot be added to or deleted**
1. Configurations are applied to a device (sensor / element) via a DeviceConfiguration
1. There can be exactly one (1) DeviceConfiguration per device
1. Creating a new DeviceConfiguration for a particular device overwrites any prior DeviceConfiguration for that device
1. DeviceConfigurations **can** be deleted

Usage:

Given any device (sensor/element), you can fetch the DeviceConfiguration with `device_configuration`

```ruby
[3] pry(main)> e = client.element("56618098-5145-445c-a6db-805eaf37ff51")
...
[4] pry(main)> device_config = e.device_configuration
=> #<Helium::DeviceConfiguration:0x007ffb71a74738
 @client=<Helium::Client @debug=false>,
 @created_at="2016-11-18T23:18:43.576921Z",
 @id="85a9bbf9-0ff7-4678-b071-03342b6c7f91",
 @loaded=nil,
 @params=
  {"attributes"=>{},
   "relationships"=>{"device"=>{"data"=>{"id"=>"56618098-5145-445c-a6db-805eaf37ff51", "type"=>"element"}}, "configuration"=>{"data"=>{"id"=>"461b1e38-6fc2-423b-974a-ae6f504c43a1", "type"=>"configuration"}}},
   "id"=>"85a9bbf9-0ff7-4678-b071-03342b6c7f91",
   "meta"=>{"loaded"=>nil, "created"=>"2016-11-18T23:18:43.576921Z"},
   "type"=>"device-configuration"},
 @type="device-configuration",
 @updated_at=nil>
```

That DeviceConfiguration can then be inspected for its device
```ruby
[8] pry(main)> device_config.device
=> #<Helium::Element:0x007ffb7311c508
 @client=<Helium::Client @debug=false>,
 @created_at="2016-04-27T18:19:06.520815Z",
 @id="56618098-5145-445c-a6db-805eaf37ff51",
 @last_seen=nil,
 @mac="redacted",
 @name="SF Office",
 @params={"attributes"=>{"name"=>"SF Office"}, "id"=>"56618098-5145-445c-a6db-805eaf37ff51", "meta"=>{"mac"=>"redacted", "created"=>"2016-04-27T18:19:06.520815Z", "last-seen"=>nil, "updated"=>"2016-11-18T01:05:34.693805Z"}, "type"=>"element"},
 @type="element",
 @updated_at="2016-11-18T01:05:34.693805Z">
```

If the device happens to be an element, a `Helium::Element` will be returned, if it is a sensor a `Helium::Sensor` will be returned

Of course all of this would be pointless if we couldn't also see the configuration
```ruby
[9] pry(main)> device_config.configuration
=> #<Helium::Configuration:0x007ffb7302ce68
 @client=<Helium::Client @debug=false>,
 @created_at="2016-11-09T21:43:03.471454Z",
 @id="461b1e38-6fc2-423b-974a-ae6f504c43a1",
 @params={"attributes"=>{"somenumber"=>"foo", "wifipasswd"=>"123456789", "jaredtest"=>true}, "relationships"=>{"device-configuration"=>{"data"=>[{"id"=>"85a9bbf9-0ff7-4678-b071-03342b6c7f91", "type"=>"device-configuration"}]}}, "id"=>"461b1e38-6fc2-423b-974a-ae6f504c43a1", "meta"=>{"created"=>"2016-11-09T21:43:03.471454Z"}, "type"=>"configuration"},
 @settings={"somenumber"=>"foo", "wifipasswd"=>"123456789", "jaredtest"=>true},
 @type="configuration",
 @updated_at=nil>
```

The settings can be fetched for any configuration with the `settings` call
```
[10] pry(main)> device_config.configuration.settings
=> {"somenumber"=>"foo", "wifipasswd"=>"123456789", "jaredtest"=>true}
```

To **create** a Configuration and DeviceConfiguration you simply apply the k/v pairs you want for the configuration

```ruby
some_hash = {"pr_docs" => true,  "some number" => 8675309 }
new_configuration = client.create_configuration(some_hash)
```

From there, to create a DeviceConfiguration you must have both a Configuration and a Device
```ruby
some_config = client.configuration("ffffffff-c7ff-4202-8426-a864d4bd9624")
some_sensor = client.sensor("ffffffff-9cda-4313-9cf7-cffee5d57050")
 new_device_configuration = client.create_device_configuration(some_sensor, some_config)
```

If you want to apply a new configuration to that sensor, once again you can just create a new DeviceConfiguration and it will overwrite the previous one.  You can also delete a DeviceConfiguration by simply calling `new_device_configuration.destroy`

### CODE REVIEW

File: lib/helium/client/configurations.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        device_confs_data = JSON.parse(response.body)["data"]
+
+        device_confs = device_confs_data.map do |dc|
+          DeviceConfiguration.new(client: self, params: dc)
+        end
+
+        return device_confs
+      end
+
+      # Configurations are immutable, so no updates are available
```
Comment (allenan) [inline lib/helium/client/configurations.rb]: maybe resource should have an immutable true/false property that prevents attempting patch calls when it's set to true? can leave that for later refactoring though

Comment (jaredmorrow) [inline lib/helium/client/configurations.rb]: Possibly.  Any PATCH or DELETE will fail on that resource

File: lib/helium/client/device_configurations.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        end
+
+        return device
+      end
+
+      def create_device_configuration(device, configuration)
+        path = "/device-configuration"
+
+        body = {
+          data: {
```
Comment (allenan) [inline lib/helium/client/device_configurations.rb]: something else we should consider refactoring at a later date: Resource#create should be able to handle these relationships in addition to the attributes

Comment (jaredmorrow) [inline lib/helium/client/device_configurations.rb]: I was on the fence about how to deal with seemingly one-off things like this.  Probably worth a refactor once we add in `metadata` support.

**Comment by allenan** [comment]
I'm +1 on this. some good refactoring opportunities for later

**Pull request merged**